pkgtools/libkver/files/lib/kver.3
.\" $NetBSD: kver.3,v 1.6 2013/07/20 21:50:53 wiz Exp $
.\"
.Dd October 31, 2012
.Os
.Dt KVER 3
.Sh NAME
.Nm kver
.Nd override system version library
.Sh SYNOPSIS
LD_PRELOAD=@PREFIX@/lib/libkver.so uname -a
.Sh DESCRIPTION
The
.Nm
library provides implementations of
.Xr sysctl 3 and
.Xr uname 3 that
shadow the real implementation and could be configured to override the
information identifying the current system.
This is useful for, among other things, building packages in a
sandbox/chrooted environment for a different
.Nx
release then the system hosting the sandbox.
.Pp
The library is usually loaded by the run-time link-editor via the
.Ev LD_PRELOAD
hook.
The
.Xr kver 8
utility provides a simple wrapper for convenience.
.Pp
For each variable that may be overridden,
.Nm
first attempts to obtain an override value from an environment variable;
if the environment variable is not set then it uses
.Xr readlink 2
to get the target of a symbolic link, and treats that as the override value;
if that also fails, then the variable is not overridden.
If none of the variables are overridden, or if an error occurs, then
the
.Nm
library emits a warning message on the standard error output,
and allows the program to run without any overrides.
.Pp
The following table lists all the variables that may be overridden,
giving the environment variable name, the symbolic link name,
and a description of the way the value is used.
.Pp
.Bl -column "LIBKVER_MACHINE_ARCH" "/libkver_machine_arch"
.It Sy Environment Ta Sy Symlink Ta Sy Description
.
.It Ev LIBKVER_OSTYPE Ta Pa /libkver_ostype Ta
The operating system type, e.g.\&
.Qq Nx .
Overrides the
.Xr sysctl 7
variable
.Va kern.ostype .
Overrides the result from
.Xr uname 1
with the
.Fl s
flag.
Overrides the
.Va sysname
element of the
.Vt "struct utsname"
result from
.Xr uname 3 .
.
.It Ev LIBKVER_OSRELEASE Ta Pa /libkver_osrelease Ta
The operating system release, e.g.\&
.Qq "6.0_STABLE" .
Overrides the
.Xr sysctl 7
variable
.Va kern.osrelease .
Overrides the result from
.Xr uname 1
with the
.Fl r
flag.
Overrides the
.Va release
element of the
.Vt "struct utsname"
result from
.Xr uname 3 .
.Pp
This value is also converted to an integer, which is used to
override the value of the
.Xr sysctl 7
variable
.Va kern.osrevision .
.
.It Ev LIBKVER_MACHINE Ta Pa /libkver_machine Ta
The operating hardware platform, e.g.\&
.Dq "amd64" .
Overrides the
.Xr sysctl 7
variable
.Va hw.machine .
Overrides the result from
.Xr uname 1
with the
.Fl m
flag.
Overrides the
.Va machine
element of the
.Vt "struct utsname"
result from
.Xr uname 3 .
.
.It Ev LIBKVER_MACHINE_ARCH Ta Pa /libkver_machine_arch Ta
The machine processor architecture, e.g.\&
.Dq "x86_64" .
Overrides the
.Xr sysctl 7
variable
.Va hw.machine_arch .
Overrides the result from
.Xr uname 1
with the
.Fl p
flag.
.El
.Sh EXAMPLES
.Dl env LD_PRELOAD=/lib/libkver.so LIBKVER_OSRELEASE=1.5 uname -r
.Pp
.Dl ln -s 1.6.2 /libkver_osrelease
.Dl env LD_PRELOAD=/lib/libkver.so uname -r
.Sh DIAGNOSTICS
.Bl -diag
.It "libkver: uname"
.Xr uname 3
failed.
.It "libkver: not configured"
None of the configurable variables has been overridden.
.It "libkver: sysctl hw.machine_arch"
.Xr sysctl 3
failed to retrieve the value of the
.Va hw.machine_arch
variable.
.It "libkver: cannot convert osrelease to osrevision: %s"
The osrelease string could not be converted to an integer for use by the
.Va kern.osrevision
.Xr sysctl 7
variable.
.Sh SEE ALSO
.Xr ld.so 1 ,
.Xr uname 1 ,
.Xr sysctl 3 ,
.Xr uname 3 ,
.Xr kver 8
.Sh AUTHORS
.An Stoned Elipot Aq Mt seb@NetBSD.org
.An Alan Barrett Aq Mt apb@NetBSD.org


doc/NEWS
# $NetBSD: NEWS,v 1.7 2012/09/30 09:02:10 asau Exp $

PKGSRC NEWS
===========

  This file contains information about important infrastructural changes in
  pkgsrc. The intended readers are the pkgsrc developers as well as anyone
  tracking pkgsrc-current.

Staged installation is default [asau 2012-08-14]
	pkgsrc has changed to use unprivileged staged installation by default.
	This makes it possible to prevent some installation mistakes.
	It also makes detection of package conflicts easier.
	Previously you had to set PKG_DESTDIR_SUPPORT to "user-destdir"
	to invoke this functionality. Set PKG_DESTDIR_SUPPORT to "none"
	to revert to direct installation.

buildlink3.mk layout changed [joerg 2009-03-20]
	The buildlink3.mk files have been reorganized to reduce
	redundant parsing for deeply nested buildlink3 trees.
	The createbuildlink package has been updated to create
	the new format.

PLIST_VARS introduced [jlam 20080413]
	PLIST_VARS is a simpler way to do conditional PLIST entries.
	The old method of doing e.g. PLIST_SUBST+= FOO="@comment " can
	now be done by unconditionally setting PLIST_VARS+=foo and
	conditionally setting PLIST.foo=yes. This will result in substitution
	of ${PLIST.foo} in the PLIST with either "" or "@comment ".

SKIP_PORTABILITY_CHECK is obsolete [rillig 20080220]
	After being deprecated for about 16 months, this early misdesign
	has finally been obsoleted and removed from pkgsrc.

Check for stripped/unstripped binaries [rillig 20080213]
	A new check has been added in mk/check/check-stripped.mk to
	assure that the installed binaries conform to the setting of the
	INSTALL_UNSTRIPPED variable. It is not enabled by default due to
	lack of testing.

OpenSSL update [tnn 20080118]
	OpenSSL in pkgsrc switched from the 0.9.7 to the 0.9.8 branch. To
	accomodate for this, 596 packages dependent on openssl had their
	revisions bumped.

pkgsrc-2007Q4 released [agc 20080105]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2008/01/05/0000.html

Simplifying meta-packages [rillig 20071219]
	The variable META_PACKAGE was introduced. This simplifies creation
	and maintenance of packages that do not install files.

Replace a package with different name [gdt 20071107]
	The variable OLDNAME was introduced for "make replace" to
	replace a package with a name other than the new one, via
	"make replace OLDNAME=oldpkgname".  Read the announcement at:
	http://mail-index.netbsd.org/tech-pkg/2007/11/07/0002.html
	
pkgsrc-2007Q3 released [agc 20071015]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/10/15/0000.html

Features framework [jlam 20070907]
	A "system features" framework has been added. It will be used to
	automatically supply missing "basic" headers and libraries for old
	systems. This is supported through the USE_FEATURES variable.

X.org 6.9 removed [joerg 20070817]
	monolithic X.org 6.9 was removed from pkgsrc. It was obsoleted by the
	modular X.org (7.x) packages. Transitioning users should remove all old
	X.org packages and set X11_TYPE=modular in mk.conf.

Emulator framework [jlam 20070729]
	A new emulator framework has been added. This will handle all
	binary-only packages that require emulation on the native OS.
	Please see pkgsrc/mk/emulator/README.

audit-packages merged with pkg_install [adrianp 20070714]
	The utility for auditing installed packages for security problems,
	audit-packages, has been rewritten in C. It is now part of the
	pkgtools/pkg_install package.
	
pkgsrc-2007Q2 released [agc 20070701]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/07/01/0000.html

New framework for bulk builds [joerg 20070619]
	pbulk, the new automated pkgsrc bulk build framework was imported.

pkgsrc-2007Q1 released [agc 20070419]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/04/19/0000.html

pkgsrc-2006Q4 released [agc 20070104]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/01/04/0000.html

modular X.org imported [joerg 20061114]
	The modular X.org packages were imported.

pkgsrc-2006Q3 released [agc 20061001]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2006/10/01/0000.html


pkgtools/libkver/files/lib/kver.c
/*	$NetBSD: kver.c,v 1.11 2012/10/31 14:41:55 apb Exp $	*/

#define sysctl _sysctl
#define uname _uname

#include <sys/types.h>
#include <sys/param.h>
#include <sys/sysctl.h>
#include <sys/utsname.h>
#include <assert.h>
#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <err.h>
#include <stdlib.h>
#include <ctype.h>
#include <unistd.h>

#undef sysctl
#undef uname


#define KVER_VERSION_FMT "%s %s (LIBKVER) #0: Tue Jan 19 00:00:00 UTC 2038 root@localhost:/sys/arch/%s/compile/LIBKVER"

#define KVER_NOT_INITIALIZED_OSRELEASE	-2
#define KVER_INVALID_OSRELEASE	-1

/*
 * How to override the operating system name; e.g. "NetBSD".
 * Used by sysctl kern.ostype, uname -s, struct utsname.sysname
 */
#ifndef PATH_LIBKVER_OSTYPE
#define PATH_LIBKVER_OSTYPE "/libkver_ostype"
#endif
#ifndef VAR_LIBKVER_OSTYPE
#define VAR_LIBKVER_OSTYPE "LIBKVER_OSTYPE"
#endif

/*
 * How to override the operating system release level; e.g. "6.0_STABLE".
 * Used by sysctl kern.osrelease, uname -r, struct utsname.release
 *
 * This is also coverted to an integer, e.g. 601000000,
 * for use by sysctl.kern.osrevision.
 */
#ifndef PATH_LIBKVER_OSRELEASE
#define PATH_LIBKVER_OSRELEASE "/libkver_osrelease"
#endif
#ifndef VAR_LIBKVER_OSRELEASE
#define VAR_LIBKVER_OSRELEASE "LIBKVER_OSRELEASE"
#endif

/*
 * How to override the machine hardware platform, e.g. "i386".
 * Used by sysctl hw.machine, uname -m, struct utsname.machine.
 */
#ifndef PATH_LIBKVER_MACHINE
#define PATH_LIBKVER_MACHINE "/libkver_machine"
#endif
#ifndef VAR_LIBKVER_MACHINE
#define VAR_LIBKVER_MACHINE "LIBKVER_MACHINE"
#endif

/*
 * How to override the machine processor architecture, e.g. "i386".
 * Used by sysctl hw.machine_arch, uname -p.
 */
#ifndef PATH_LIBKVER_MACHINE_ARCH
#define PATH_LIBKVER_MACHINE_ARCH "/libkver_machine_arch"
#endif
#ifndef VAR_LIBKVER_MACHINE_ARCH
#define VAR_LIBKVER_MACHINE_ARCH "LIBKVER_MACHINE_ARCH"
#endif

static struct kver {
	char ostype[_SYS_NMLN];
	char osrelease[_SYS_NMLN];
	int osrevision;
	char version[_SYS_NMLN];
	char machine[_SYS_NMLN];
	char machine_arch[_SYS_NMLN];
}    kver = {
	.osrevision = KVER_NOT_INITIALIZED_OSRELEASE
};

static struct utsname real_utsname;

#define KVER_NOT_INITIALIZED	\
	(kver.osrevision == KVER_NOT_INITIALIZED_OSRELEASE)

#define KVER_BADLY_INITIALIZED	(kver.osrevision == KVER_INVALID_OSRELEASE)

#define SYSCTL_STRING(oldp, oldlenp, str)				\
	if (oldlenp) {							\
		size_t len = strlen(str) + 1;				\
		if (!oldp)						\
			*oldlenp = len;					\
		else {							\
			if (*oldlenp < len) {				\
				r = ENOMEM;				\
				len = *oldlenp;				\
			} else						\
				*oldlenp = len;				\
			(void) strncpy(oldp, str, len);			\
		}							\
	}

#if __NetBSD_Version__ >= 399001600
#define SYSCTL_CONST	const
#else
#define SYSCTL_CONST
#endif

static int
str2osrevision(char *s)
{
	char c;
	int n, r = 0;

	if (s == NULL || *s == 0)
		return KVER_INVALID_OSRELEASE;

	if (!isdigit(*s))
		return KVER_INVALID_OSRELEASE;

	/* first digit: major */
	for (n = 0; isdigit(*s); s++) {
		n = (n * 10) + (*s - '0');
	}
	if (*s == 0 || *s != '.')
		return KVER_INVALID_OSRELEASE;
	r += (n * 100000000);

	/* second digit: minor */
	for (s++, n = 0; isdigit(*s); s++) {
		n = (n * 10) + (*s - '0');
	}
	r += (n * 1000000);

	/* nothing more, return */
	if (*s == 0)
		return r;

	/* optional third digit: patchlevel */
	if (*s == '.') {
		for (s++, n = 0; isdigit(*s); s++) {
			n = (n * 10) + (*s - '0');
		}
		r += (n * 100);
	}
	/* nothing more, return */
	if (*s == 0)
		return r;
	/* or optional underscore followed by release status */
	if (*s == '_')
		/* ignore the trailing label */
		return r;
	/* or optional letters: release */
	n = 0;
	c = 'Z';
	while (*s >= 'A' && *s <= 'Z') {
		if (c != 'Z' && *s != 0)
			break;
		c = *s;
		n += *s++ - '@';
	}
	if (n > 99)
		return KVER_INVALID_OSRELEASE;
	if (*s == 0)
		return r + (n * 10000);

	/* return on error if we end up here */
	return KVER_INVALID_OSRELEASE;
}

/*
 * Initialise one element of struct kver.
 *
 * The result is based on getenv(envvarname), or readlink(linkname),
 * or defaultval.  If the provided defaultval is NULL, then the empty
 * string is used.
 *
 * Returns 1 if a non-default value was used, 0 if a default was used.
 */
static int
kver_init_var(char *dst, size_t dstlen, const char *envvar,
    const char *linkname, const char *defaultval)
{
	ssize_t len;
	char *v;

	v = getenv(envvar);
	if (v) {
		(void) strlcpy(dst, v, dstlen);
		return 1;
	}
	len = readlink(linkname, dst, dstlen);
	if (len > 0) {
		dst[len] = '\0';
		return 1;
	}
	if (defaultval) {
		(void) strlcpy(dst, defaultval, dstlen);
		return 0;
	}
	dst[0] = '\0';
	return 0;
}

static void
kver_initialize(void)
{
	char *v;
	int i;
	int nok = 0;
	char buf[PATH_MAX + 1];
	kver.osrevision = KVER_INVALID_OSRELEASE;	/* init done */

	if (_uname(&real_utsname) != 0) {
		warn("libkver: uname");
		return;
	}

	/*
	 * For each variable that can be overridden: try the
	 * environment variable, then try the symlink, or fall back
	 * to the unmodified (existing) value.
	 */
	nok += kver_init_var(kver.ostype, sizeof(kver.ostype),
	    VAR_LIBKVER_OSTYPE, PATH_LIBKVER_OSTYPE,
	    real_utsname.sysname);
	nok += kver_init_var(kver.osrelease, sizeof(kver.osrelease),
	    VAR_LIBKVER_OSRELEASE, PATH_LIBKVER_OSRELEASE,
	    real_utsname.release);
	nok += kver_init_var(kver.machine, sizeof(kver.machine),
	    VAR_LIBKVER_MACHINE, PATH_LIBKVER_MACHINE,
	    real_utsname.machine);
	nok += kver_init_var(kver.machine_arch, sizeof(kver.machine_arch),
	    VAR_LIBKVER_MACHINE_ARCH, PATH_LIBKVER_MACHINE_ARCH,
	    "");

	/*
	 * warning if the default was used for all variables.  no warning
	 * if at least one non-default value was used.
	 */
	if (nok == 0) {
		warnx("libkver: not configured");
		return;
	}

	/* machine_arch is not in struct utsname, so get default from sysctl */
	if (kver.machine_arch[0] == '\0') {
		int mib[] = { CTL_HW, HW_MACHINE_ARCH, 0 };
		int len = sizeof(kver.machine_arch);
		int err;

		err = _sysctl(mib, 2, kver.machine_arch, &len, NULL, 0);
		if (err < 0) {
			warn("libkver: sysctl hw.machine_arch");
			return;
		}
	}

	/* combine several strings to create kver.version */
	(void) snprintf(kver.version, sizeof(kver.version), KVER_VERSION_FMT,
	    kver.ostype, kver.osrelease, kver.machine);

	/*
	 * Convert string osrelease to integer osrevision.
	 * This must be the last thing we do, so that any failure
	 * resuls in kver.osrelease being set to KVER_INVALID_OSRELEASE.
	 */
	kver.osrevision = str2osrevision(kver.osrelease);
	if (kver.osrevision == KVER_INVALID_OSRELEASE) {
		warnx("libkver: cannot convert osrelease to osrevision: %s",
		    kver.osrelease);
	}

	return;
}

int
sysctl(SYSCTL_CONST int *name, u_int namelen, void *oldp, size_t * oldlenp,
    const void *newp, size_t newlen)
{
	int r = 0;

	_DIAGASSERT(name != NULL);

	if (newp != (void *) NULL)
		goto real;

	if (KVER_NOT_INITIALIZED)
		kver_initialize();

	if (KVER_BADLY_INITIALIZED || namelen != 2)
		goto real;

	switch (name[0]) {
	case CTL_KERN: {
		switch (name[1]) {
		case KERN_OSTYPE:
			SYSCTL_STRING(oldp, oldlenp, kver.ostype);
			return (r);
		case KERN_OSRELEASE:
			SYSCTL_STRING(oldp, oldlenp, kver.osrelease);
			return (r);
		case KERN_OSREV:
			if (oldlenp) {
				if (!oldp)
					*oldlenp = sizeof(int);
				else {
					if (*oldlenp < sizeof(int))
						return (ENOMEM);
					*oldlenp = sizeof(int);
					*((int *) oldp) = kver.osrevision;
				}
			}
			return (r);
		case KERN_VERSION:
			SYSCTL_STRING(oldp, oldlenp, kver.version);
			return (r);
		}
	}
	case CTL_HW: {
		switch (name[1]) {
		case HW_MACHINE:
			SYSCTL_STRING(oldp, oldlenp, kver.machine);
			return (r);
		case HW_MACHINE_ARCH:
			SYSCTL_STRING(oldp, oldlenp, kver.machine_arch);
			return (r);
		}
	}
	}
real:	return (_sysctl(name, namelen, oldp, oldlenp, newp, newlen));
}

int
uname(struct utsname * n)
{
	if (KVER_NOT_INITIALIZED)
		kver_initialize();

	if (KVER_BADLY_INITIALIZED)
		return _uname(n);

	(void) strncpy(n->sysname, kver.ostype, _SYS_NMLN);
	(void) strncpy(n->nodename, real_utsname.nodename, _SYS_NMLN);
	(void) strncpy(n->release, kver.osrelease, _SYS_NMLN);
	(void) strncpy(n->version, kver.version, _SYS_NMLN);
	(void) strncpy(n->machine, kver.machine, _SYS_NMLN);
	return 0;
}
